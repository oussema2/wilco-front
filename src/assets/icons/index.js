/* eslint-disable global-require */
export const aircraft = require( './aircraft.png' );
export const airport = require( './airport.png' );
export const arrowLeft = require( './arrowLeft.png' );
export const arrowRight = require( './arrowRight.png' );
export const award = require( './award.png' );
export const ban = require( './ban.png' );
export const bell = require( './bell.png' );
export const binoculars = require( './binoculars.png' );
export const brandSymbol = require( './brandSymbol.png' );
export const chevronLeft = require( './chevronLeft.png' );
export const chevronRight = require( './chevronRight.png' );
export const cloud = require( './cloud.png' );
export const certificate = require( './certificate.png' );
export const chatComment = require( './chatComment.png' );
export const chatNewMessage = require( './chatNewMessage.png' );
export const checkbox = require( './checkbox.png' );
export const checkboxUnselected = require( './checkboxUnselected.png' );
export const comment = require( './comment.png' );
export const cross = require( './cross.png' );
export const eye = require( './eye.png' );
export const eyeSlash = require( './eyeSlash.png' );
export const exclamationCircle = require( './exclamationCircle.png' );
export const exclamationTriangle = require( './exclamationTriangle.png' );
export const filter = require( './filter.png' );
export const flightInfoAircraft = require( './flightInfoAircraft.png' );
export const home = require( './home.png' );
export const horizontalLogo = require( './horizontalLogo.png' );
export const infoCircle = require( './infoCircle.png' );
export const logOut = require( './logOut.png' );
export const map = require( './map.png' );
export const meatballsMenu = require( './meatballsMenu.png' );
export const megaphone = require( './megaphone.png' );
export const onBoardingFirst = require( './onBoardingFirst.png' );
export const onBoardingSecond = require( './onBoardingSecond.png' );
export const onBoardingThird = require( './onBoardingThird.png' );
export const planeDeparture = require( './planeDeparture.png' );
export const plus = require( './plus.png' );
export const photoVideo = require( './photoVideo.png' );
export const privacyOnlyMe = require( './privacyOnlyMe.png' );
export const privacyPublic = require( './privacyPublic.png' );
export const remove = require( './remove.png' );
export const route = require( './route.png' );
export const rulerVertical = require( './rulerVertical.png' );
export const star = require( './star.png' );
export const starMyFeed = require( './starMyFeed.png' );
export const search = require( './search.png' );
export const settings = require( './settings.png' );
export const sliders = require( './sliders.png' );
export const speed = require( './speed.png' );
export const stopwatch = require( './stopwatch.png' );
export const thumbsUp = require( './thumbsUp.png' );
export const user = require( './user.png' );
export const users = require( './users.png' );
export const vector = require( './vector.png' );
