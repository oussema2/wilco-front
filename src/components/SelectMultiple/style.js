export function mergeStyles( ...args ) {
	return Array.prototype.concat.apply( [], args );
}
