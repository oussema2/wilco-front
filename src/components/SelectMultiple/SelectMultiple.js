import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
	View, Text, Image, Pressable
} from 'react-native';
import styles from './SelectMultiple.styles';
import { mergeStyles } from './style';

const itemType = PropTypes.oneOfType( [
	PropTypes.string,
	PropTypes.shape( { label: PropTypes.any, value: PropTypes.any } )
] );

const styleType = PropTypes.oneOfType( [
	PropTypes.object,
	PropTypes.number,
	PropTypes.array
] );

const sourceType = PropTypes.oneOfType( [ PropTypes.object, PropTypes.number ] );

export default class SelectMultiple extends Component {
	// eslint-disable-next-line react/static-property-placement
	static propTypes = {
		testID: PropTypes.string,
		items: PropTypes.arrayOf( itemType ).isRequired,
		selectedItems: PropTypes.arrayOf( itemType ),
		maxSelect: PropTypes.number,
		onSelectionsChange: PropTypes.func.isRequired,
		checkboxSource: sourceType.isRequired,
		selectedCheckboxSource: sourceType.isRequired,
		renderLabel: PropTypes.func,
		rowStyle: styleType,
		checkboxStyle: styleType,
		labelStyle: styleType,
		selectedRowStyle: styleType,
		selectedCheckboxStyle: styleType,
		selectedLabelStyle: styleType,
		ParentComponent: PropTypes.elementType
	}

	// eslint-disable-next-line react/static-property-placement
	static defaultProps = {
		testID: 'select-multiple-testID',
		selectedItems: [],
		rowStyle: {},
		checkboxStyle: {},
		labelStyle: {},
		maxSelect: null,
		renderLabel: null,
		selectedRowStyle: null,
		selectedCheckboxStyle: null,
		selectedLabelStyle: null,
		ParentComponent: View
	}

	items = undefined

	selectedItems = undefined

	rowData = []

	onRowPress( row ) {
		const { label, value } = row;
		let { onSelectionsChange, selectedItems, maxSelect } = this.props;

		selectedItems = ( selectedItems || [] ).map( this.toLabelValueObject );

		const index = selectedItems.findIndex( ( selectedItem ) => selectedItem.value === value );

		if ( index > -1 ) {
			selectedItems = selectedItems.filter( ( selectedItem ) => selectedItem.value !== value );
		} else if ( maxSelect != null && selectedItems.length >= maxSelect ) {
			return;
		} else {
			selectedItems = selectedItems.concat( { label, value } );
		}

		onSelectionsChange( selectedItems, { label, value } );
	}

	getRowData( items, selectedItems ) {
		if ( items !== this.items || selectedItems !== this.selectedItems ) {
			this.items = items;
			this.selectedItems = selectedItems;
			this.rowData = items?.map( this.toLabelValueObject );
			selectedItems = ( selectedItems || [] ).map( this.toLabelValueObject );

			this.rowData?.forEach( ( item ) => {
				item.selected = selectedItems.some( ( i ) => i.value === item.value );
			} );
		}

		return this.rowData;
	}

	keyExtractor = ( item, index ) => index.toString()

	toLabelValueObject( obj ) {
		if ( Object.prototype.toString.call( obj ) === '[object String]' ) {
			return { label: obj, value: obj };
		}
		return { label: obj.label, value: obj.value };
	}

	renderLabel = ( label, style, selected ) => {
		const { renderLabel } = this.props;
		if ( renderLabel ) {
			return renderLabel( label, style, selected );
		}
		return (
			<Text numberOfLines={1} style={style}>{label}</Text>
		);
	}

	renderItemRow = ( item, index ) => {
		let {
			checkboxSource,
			rowStyle,
			labelStyle,
			checkboxStyle
		} = this.props;

		const {
			selectedCheckboxSource,
			selectedRowStyle,
			selectedCheckboxStyle,
			selectedLabelStyle
		} = this.props;

		if ( item.selected ) {
			checkboxSource = selectedCheckboxSource;
			rowStyle = mergeStyles( styles.row, rowStyle, selectedRowStyle );
			checkboxStyle = mergeStyles( styles.checkbox, checkboxStyle, selectedCheckboxStyle );
			labelStyle = mergeStyles( styles.label, labelStyle, selectedLabelStyle );
		} else {
			rowStyle = mergeStyles( styles.row, rowStyle );
			checkboxStyle = mergeStyles( styles.checkbox, checkboxStyle );
			labelStyle = mergeStyles( styles.label, labelStyle );
		}

		return (
			<Pressable testID={`row-${item.value}-testID`} onPress={() => this.onRowPress( item )} key={index}>
				<View style={rowStyle}>
					<Image style={checkboxStyle} source={checkboxSource} />
					<View style={styles.labelContainer}>
						{this.renderLabel( item.label, labelStyle, item.selected )}
					</View>
				</View>
			</Pressable>
		);
	}

	render() {
		const {
			items, selectedItems, ParentComponent, testID
		} = this.props;
		return (
			<ParentComponent testID={testID}>
				{this.getRowData( items, selectedItems )?.map(
					( item, index ) => this.renderItemRow( item, index ) )}
			</ParentComponent>
		);
	}
}
