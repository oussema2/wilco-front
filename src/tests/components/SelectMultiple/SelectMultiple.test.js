import * as React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import SelectMultiple from '../../../components/SelectMultiple/SelectMultiple';
import { checkbox, checkboxUnselected } from '../../../assets/icons';

describe( 'SelectMultiple', () => {
	describe( 'Autolink component with the default props', () => {
		it( 'renders the SelectMultiple component correctly', async () => {
			const onSelectionsChange = jest.fn();
			const items = [ { value: 1, label: 'label 1' } ];
			const component = render( <SelectMultiple
				items={items}
				selectedCheckboxSource={checkbox}
				checkboxSource={checkboxUnselected}
				onSelectionsChange={onSelectionsChange}
			/> );

			expect( component.queryByTestId( 'select-multiple-testID' ) ).not.toBeNull();
			expect( component ).toMatchSnapshot();
		} );
	} );

	describe( 'with custom testID prop', () => {
		it( 'uses the given testID', () => {
			const items = [ { value: 1, label: 'label 1' } ];
			const testID = 'test-testID';
			const component = render( <SelectMultiple
				testID={testID}
				items={items}
				selectedCheckboxSource={checkbox}
				checkboxSource={checkboxUnselected}
			/> );

			expect( component.queryByTestId( testID ) ).toBeDefined();
		} );
	} );

	describe( 'with custom onSelectionsChange prop', () => {
		it( 'calls to onSelectionsChange function', () => {
			const onSelectionsChange = jest.fn();
			const items = [ { value: 1, label: 'label 1' } ];
			const testID = 'test-testID';
			const component = render( <SelectMultiple
				testID={testID}
				items={items}
				selectedCheckboxSource={checkbox}
				checkboxSource={checkboxUnselected}
				onSelectionsChange={onSelectionsChange}
			/> );

			fireEvent.press( component.getByTestId( 'row-1-testID' ) );
			expect( onSelectionsChange ).toHaveBeenCalledTimes( 1 );
		} );
	} );
} );
