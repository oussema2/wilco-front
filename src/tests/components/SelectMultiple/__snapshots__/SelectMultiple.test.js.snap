// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectMultiple Autolink component with the default props renders the SelectMultiple component correctly 1`] = `
<View
  testID="select-multiple-testID"
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="row-1-testID"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderBottomColor": "#cccccc",
            "borderBottomWidth": 1,
            "flex": 1,
            "flexDirection": "row",
            "padding": 15,
          },
          Object {},
        ]
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/assets/icons/checkboxUnselected.png",
          }
        }
        style={
          Array [
            Object {
              "height": 24,
              "marginRight": 5,
              "width": 24,
            },
            Object {},
          ]
        }
      />
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {},
              Object {},
            ]
          }
        >
          label 1
        </Text>
      </View>
    </View>
  </View>
</View>
`;
